packages/go-gen/src/go.rs

```diff
@@ -0,0 +1,235 @@
+        // generate type
+        writeln!(f, "type {} struct {{", self.name)?;
+        // generate fields
+        {
+            let mut f = indented(f);
+            for field in &self.fields {
+                writeln!(f, "{}", field)?;
+            }
+        }
+        f.write_char('}')?;
+        Ok(())
+    }
+}
+
+pub struct GoField {
+    /// The name of the field in Rust (snake_case)
+    pub rust_name: String,
+    /// The documentation of the field
+    pub docs: Option<String>,
+    /// The type of the field
+    pub ty: GoType,
+}
+
+impl Display for GoField {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // documentation
+        format_docs(f, self.docs.as_deref())?;
+        // {field} {type} `json:"{field}"`
+        write!(
+            f,
+            "{} {} `json:\"{}",
+            self.rust_name.to_pascal_case(),
+            self.ty,
+            self.rust_name
+        )?;
+        if self.ty.is_nullable {
+            f.write_str(",omitempty")?;
+        }
+        f.write_str("\"`")
+    }
+}
+
+pub struct GoType {
+    /// The name of the type in Go
+    pub name: String,
+    /// Whether the type should be nullable
+    /// This will add `omitempty` to the json tag and use a pointer type if
+    /// the type is not a basic type
+    pub is_nullable: bool,
+}
+
+impl GoType {
+    pub fn is_basic_type(&self) -> bool {
+        const BASIC_GO_TYPES: &[&str] = &[
+            "string",
+            "bool",
+            "int",
+            "int8",
+            "int16",
+            "int32",
+            "int64",
+            "uint",
+            "uint8",
+            "uint16",
+            "uint32",
+            "uint64",
+            "float32",
+            "float64",
+            "byte",
+            "rune",
+            "uintptr",
+            "complex64",
+            "complex128",
+        ];
+        BASIC_GO_TYPES.contains(&&*self.name)
+    }
+}
+
+impl Display for GoType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.is_nullable && !self.is_basic_type() {
+            // if the type is nullable and not a basic type, use a pointer
+            f.write_char('*')?;
+        }
+        f.write_str(&self.name)
+    }
+}
+
+fn format_docs(f: &mut fmt::Formatter, docs: Option<&str>) -> fmt::Result {
+    if let Some(docs) = docs {
+        for line in docs.lines() {
+            f.write_str("// ")?;
+            f.write_str(line)?;
+            f.write_char('\n')?;
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn go_type_display_works() {
+        let ty = GoType {
+            name: "string".to_string(),
+            is_nullable: true,
+        };
+        let ty2 = GoType {
+            name: "string".to_string(),
+            is_nullable: false,
+        };
+        assert_eq!(format!("{}", ty), "string");
+        assert_eq!(format!("{}", ty2), "string");
+
```

Comment:
Could you also test binary field (`Binary` in Rust, `[]byte` in Go)?

Reply:
That's tested in [main.rs](https://github.com/CosmWasm/cosmwasm/blob/591b198a1ec593a515b373356437ac7c4f749eeb/packages/go-gen/src/main.rs#L215), since the mapping does not happen in go.rs.